### PR TITLE
Smooth fading of animations in and out.

### PR DESCRIPTION
**Pull request by howard-stearns**
Also, turn on the secret rig animations from Javascript with MyAvatar.setEnableRigAnimations(true). (persists)

BTW: In this form, you probably want Avatar -> Enable avatar collision whenever setEnableRigAnimations is true -- unless you WANT to see continuous walking.

### CODE REVIEW

**Comment by howard-stearns** [comment]
I'll try to resolve the conflicts on the train... Sigh.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 61198a658c3a43da82e2bb0264b958cb5bd87e0d available at: https://gist.github.com/ad5e3682d09e2924fd57<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 731bcba9563103fa32d812b0acfd167855c2698d available at: https://gist.github.com/ad5e3682d09e2924fd57<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a1f86cc7b8d7de6b158a731d6d53e2cad37aa774 available at: https://gist.github.com/ad5e3682d09e2924fd57<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.dmg

File: libraries/animation/src/AnimationHandle.h [hunk 1]
```diff
@@ -64,6 +64,10 @@ class AnimationHandle : public QObject, public std::enable_shared_from_this<Anim
     void setPriority(float priority);
     float getPriority() const { return _priority; }
     void setMix(float mix) { _mix = mix; }
+    void setFade(float fade) { _fade = fade; }
+    float getFade() { return _fade; }
+    void setFadePerSecond(float fadePerSecond) { _fadePerSecond = fadePerSecond; }
+    float getFadePerSecond() { return _fadePerSecond; }
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimationHandle.h]: getFade*() should be const methods.

File: libraries/animation/src/Rig.cpp [hunk 2]
```diff
@@ -428,13 +430,57 @@ void Rig::computeMotionAnimationState(float deltaTime, const glm::vec3& worldPos
...
+    int animationIndex = 0;
+    foreach (const AnimationHandlePointer& handle, _runningAnimations) {
+        float fade = handle->getFade();
+        normalizedFades[animationIndex++] = fade;
+        fadeSum += fade;
+    }
+    // normalize our copy of the fade data
+    for (int i = 0; i < _runningAnimations.count(); i++) {
+        normalizedFades[i] /= fadeSum;
+    }
```
Comment (AndrewMeadows) [inline libraries/animation/src/Rig.cpp]: This loop is technically not required. You could perform this normalization in the subsequent loop... but you wouldn't be able to recycle the **fadeSum** variable.  You'd have to declare a new one.

File: libraries/animation/src/Rig.cpp [hunk 3]
```diff
@@ -131,16 +131,18 @@ void Rig::addAnimationByRole(const QString& role, const QString& url, float fps,
     if (startAutomatically) {
         handle->start();
     }
+    return handle;
 }
 void Rig::startAnimationByRole(const QString& role, const QString& url, float fps, float priority,
                                bool loop, bool hold, float firstFrame, float lastFrame, const QStringList& maskedJoints) {
-    addAnimationByRole(role, url, fps, priority, loop, hold, firstFrame, lastFrame, maskedJoints, true);
+    AnimationHandlePointer handle = addAnimationByRole(role, url, fps, priority, loop, hold, firstFrame, lastFrame, maskedJoints, true);
+    handle->setFadePerSecond(1.0f); // For now. Could be individualized later.
```
Comment (hyperlogic) [inline libraries/animation/src/Rig.cpp]: 1.0 is kind of a long default for fading. IMHO. 0.5f might be better.

File: libraries/animation/src/Rig.cpp [hunk 4]
```diff
@@ -428,13 +430,57 @@ void Rig::computeMotionAnimationState(float deltaTime, const glm::vec3& worldPos
...
+    foreach (const AnimationHandlePointer& handle, _runningAnimations) {
+        handle->setPriority(1.0f);
+        float normalizedFade = normalizedFades[animationIndex++];
+        // simulate() will blend each animation result into the result so far, based on the pairwise mix at at each step.
+        // i.e., slerp the 'mix' distance from the result so far towards this iteration's animation result.
+        // The formula here for mix is based on the idea that, at each step:
+        // fadeSum is to normalizedFade, as (1 - mix) is to mix
+        // i.e., fadeSum/normalizedFade = (1 - mix)/mix
+        // Then we solve for mix.
+        // Sanity check: For the first animation, fadeSum = 0, and the mix will always be 1.
```
Comment (hyperlogic) [inline libraries/animation/src/Rig.cpp]: Perhaps turn these sanity checks into actual asserts?

**Comment by hyperlogic** [comment]
:+1:  Looks good to me.

File: libraries/animation/src/AnimationHandle.h (hunk 1 above)
Comment (howard-stearns) [inline libraries/animation/src/AnimationHandle.h]: Recording discussion: we'll add the const, but the clamping will remain in the caller not the setter.

File: libraries/animation/src/Rig.cpp (hunk 2 above)
Comment (howard-stearns) [inline libraries/animation/src/Rig.cpp]: Ooh, I like that much better. Will change. Thanks for seeing that.

File: libraries/animation/src/Rig.cpp (hunk 4 above)
Comment (howard-stearns) [inline libraries/animation/src/Rig.cpp]: After re-arranging the loop, I can't figure out how to turn these particular sanity checks into asserts. However, it's still a great idea to assert that the mix is between 0 and 1, so I'm adding that.

File: libraries/animation/src/Rig.cpp (hunk 3 above)
Comment (howard-stearns) [inline libraries/animation/src/Rig.cpp]: Yes! We'll make the fadePerSecond be faster after we've had some experience and the chance to see what is really happening. Future PR.

**Comment by hifi-gustavo** [comment]
Build console output for commit: bb974edf7f5d74d2bb8a418e869f727fffb8b0ec available at: https://gist.github.com/ad5e3682d09e2924fd57<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5506.dmg

**Comment by hyperlogic** [comment]
:+1:

**Pull request merged**